src/p11_child/p11_child_openssl.c

```diff
@@ -42,8 +43,325 @@ struct p11_ctx {
+            goto err;
+        }
+        if (rv == 0) {
+            DEBUG(SSSDBG_OP_FAILURE, "Timeout on request\n");
+            break;
+        }
+        if (rv == -1) {
+            DEBUG(SSSDBG_OP_FAILURE, "Select error\n");
+            break;
+        }
+
+    }
+ err:
+    OCSP_REQ_CTX_free(ctx);
+
+    return rsp;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#define TLS_client_method SSLv23_client_method
+#define OCSP_REQUEST_add0_id OCSP_request_add0_id
+#endif
+
+OCSP_RESPONSE *process_responder(OCSP_REQUEST *req,
+                                 const char *host, const char *path,
+                                 const char *port, int use_ssl,
+                                 int req_timeout)
+{
+    BIO *cbio = NULL;
+    SSL_CTX *ctx = NULL;
+    OCSP_RESPONSE *resp = NULL;
+
+    cbio = BIO_new_connect(host);
+    if (cbio == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Error creating connect BIO\n");
+        goto end;
+    }
+    if (port != NULL)
+        BIO_set_conn_port(cbio, port);
+    if (use_ssl == 1) {
+        BIO *sbio;
+        ctx = SSL_CTX_new(TLS_client_method());
+        if (ctx == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Error creating SSL context.\n");
+            goto end;
+        }
+        SSL_CTX_set_mode(ctx, SSL_MODE_AUTO_RETRY);
+        sbio = BIO_new_ssl(ctx, 1);
+        cbio = BIO_push(sbio, cbio);
+    }
+
+    resp = query_responder(cbio, host, path, req, req_timeout);
+    if (resp == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Error querying OCSP responder\n");
+    }
+
+ end:
+    BIO_free_all(cbio);
+    SSL_CTX_free(ctx);
+    return resp;
+}
+
+static errno_t do_ocsp(struct p11_ctx *p11_ctx, X509 *cert)
+{
+    OCSP_REQUEST *ocsp_req = NULL;
+    OCSP_RESPONSE *ocsp_resp = NULL;
+    OCSP_BASICRESP *ocsp_basic = NULL;
+    OCSP_CERTID *cid = NULL;
+    STACK_OF(OPENSSL_STRING) *ocsp_urls = NULL;
+    char *url_str;
+    X509 *issuer = NULL;
+    int req_timeout = -1;
+    int status;
+    int ret = EIO;
+    int reason;
+    ASN1_GENERALIZEDTIME *revtime;
+    ASN1_GENERALIZEDTIME *thisupd;
+    ASN1_GENERALIZEDTIME *nextupd;
+    long grace_time = (5 * 60); /* Allow 5 minutes time difference when
+                                 * checking the validity of the OCSP response */
+    char *host = NULL;
+    char *path = NULL;
+    char *port = NULL;
+    int use_ssl;
+    X509_NAME *issuer_name = NULL;
+    X509_OBJECT *x509_obj;
+
+    ocsp_urls = X509_get1_ocsp(cert);
+    if (ocsp_urls == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "No OCSP URL in certificate, skipping OCSP check.\n");
+        return EOK;
+    }
+
+    if (sk_OPENSSL_STRING_num(ocsp_urls) > 1) {
+        DEBUG(SSSDBG_CONF_SETTINGS,
+              "Found more than 1 OCSP URLs, just using the first.\n");
+    }
+
+    if (p11_ctx->cert_verify_opts->ocsp_default_responder != NULL) {
+        url_str = p11_ctx->cert_verify_opts->ocsp_default_responder;
```

Comment:
Is it correct to skip the OCSP validation if the certificate has no OCSP URI defined in itself, but the ocsp_default_responder option is set?